File: src/main/java/world/Chunk.java
package world;

import block.EnumRenderLayer;
import block.TileState;
import gl.ShaderProgram;
import org.lwjgl.system.MemoryStack;
import util.FloatList;
import worker.RenderPool;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Chunk {
    private Region[] _regions = new Region[8];
    public boolean hasGenerated = false;
    public boolean hasDecorated = false;
    public boolean isLoaded = false;
    public boolean dirty = false;

    private World _world;
    public int cX;
    public int cZ;

    public Chunk(World world, int cX, int cZ) {
        _world = world;
        this.cX = cX;
        this.cZ = cZ;

        for (int i = 0;i < _regions.length;i++) {
            _regions[i] = new Region(cX, i, cZ);
        }
    }


    public TileState getTileAt(int x, int y, int z) {
        if (x > 15 || x < 0 || z > 15 || z < 0 || y < 0 || y > 127)
            return null;

        int r = y >> 4;
        if (_regions.length >= r) {
            return _regions[r].getTileAt(x, y & 15, z);
        }
        return null;
    }



    public void generate(World world) {
        world.worldType.gen.generate(world, cX, cZ);
        hasGenerated = true;
        dirty = true;
    }

    public void decorate(World world) {
        world.worldType.gen.decorate(world, cX, cZ);
        hasDecorated = true;
        dirty = true;
    }

    public List<FloatList> buildRenderData(World world, EnumRenderLayer l) {
        List<FloatList> verts = new ArrayList<>();
        for (Region region : _regions) {
            FloatList v = region.buildRenderData(world, cX, cZ, l);
            verts.add(v);
        }
        return verts;
    }

    public void draw(MemoryStack stack, EnumRenderLayer l, int uniTrans, float timer) {
        for (Region r : _regions) {
            r.draw(stack, l, uniTrans, timer);
        }
    }

    public void setTileAt(int x, int y, int z, int tile) {
        if (x > 15 || x < 0 || z > 15 || z < 0 || y < 0 || y > 127) {
            return;
        }
        int r = y >> 4;
        if (_regions.length >= r) {
            _regions[r].setTileAt(x, y & 15, z, tile);
        }
        if(this.hasGenerated && !this.dirty) {
            this.dirty = true;
            RenderPool.enqueueChunkRender(_world, this, false);
        }
    }

    public void bindRenderData(List<FloatList> verts, EnumRenderLayer l, ShaderProgram prog) {
        for(int i = 0;i < _regions.length;i++) {
            _regions[i].bindData(verts.get(i), l, prog);
        }
    }

    public void tick(World world, Random random) {
        for (Region r : _regions) {
            r.tick(world, random);
        }
    }
}

File: src/main/kotlin/world/Chunk.kt
package world

import block.EnumRenderLayer
import block.TileState
import gl.ShaderProgram
import kotlinx.coroutines.runBlocking
import util.FloatList

class Chunk(val world: World, val cX: Int, val cZ: Int) {
    private var _regions: Array<Region?> = Array(8) { null }
    var hasGenerated = false
    var hasDecorated = false
    var isLoaded = false
    var dirty = false


    fun getTileAt(x: Int, y: Int, z: Int): TileState? {
        if (x > 15 || x < 0 || z > 15 || z < 0 || y < 0 || y > 127) return null

        val r = y shr 4
        if (_regions.size >= r) {
            return _regions[r]!!.getTileAt(x, y and 15, z)
        }
        return null
    }

    init {
        for (i in 0 until _regions.size) {
            _regions[i] = Region(cX, i, cZ)
        }
    }

    fun generate(world: World) {
        world.worldType.gen.generate(world,  cX, cZ)
        hasGenerated = true
    }

    fun decorate(world: World) {
        world.worldType.gen.decorate(world, cX, cZ)
        hasDecorated = true
    }

    fun buildRenderData(world: World, l: EnumRenderLayer): MutableList<FloatList> {
        val verts = mutableListOf<FloatList>()
        for (i in 0 until _regions.size) {
            val v = _regions[i]!!.buildRenderData(world, this.cX, this.cZ, l)
            verts.add(v)
        }
        return verts
    }

    fun draw(l: EnumRenderLayer, uniTrans: Int, timer: Float) {
        for (i in 0 until _regions.size) {
            _regions[i]!!.draw(l, uniTrans, timer)
        }
    }

    fun setTileAt(x: Int, y: Int, z: Int, tile: Int) {
        if (x > 15 || x < 0 || z > 15 || z < 0 || y < 0 || y > 127) {
            return
        }
        val r = y shr 4
        if (_regions.size >= r) {
           _regions[r]!!.setTileAt(x, y and 15, z, tile)
        }
        if(this.hasGenerated && !this.dirty) {
            this.dirty = true
            val c = this
            runBlocking {
                world.renderChunkMutex.lock()
                world.renderChunkQueue.offer(RenderChunkBatch(c, false))
                world.renderChunkMutex.unlock()
            }
        }
    }

    fun bindRenderData(verts: MutableList<FloatList>, layer: EnumRenderLayer, prog: ShaderProgram) {
        for (i in 0 until _regions.size) {
            _regions[i]!!.bindData(verts[i], layer, prog)
        }
    }

    fun tick(world: World) {
        for (i in 0 until _regions.size) {
            _regions[i]!!.tick(world)
        }
    }
}